src/lib/components/availability/SetAvailabilityModal.svelte

```diff
@@ -0,0 +1,170 @@
+	let endTimeString: string = getTimeString(availability?.endTime ?? rehearsal.endTime)
+	let startTimeInvalid: boolean = false
+	let endTimeInvalid: boolean = false
+
+	/**
+	 * Submit or update the availability in the database.
+	 * @returns `true` if the mutation was successful, `false` otherwise.
+	 */
+	async function saveAvailability(): Promise<boolean> {
+		// Time input validation
+		const timeRegex = /^([0-1]?[0-9]|2[0-3]):[0-5][0-9]$/
+		startTimeInvalid = !timeRegex.test(startTimeString)
+		endTimeInvalid = !timeRegex.test(endTimeString)
+		if (startTimeInvalid || endTimeInvalid) return false
+
+		// If the participant is not available, they are unavailable for the entire rehearsal
+		if (!available) {
+			startTimeString = getTimeString(rehearsal.startTime)
+			endTimeString = getTimeString(rehearsal.endTime)
+		}
+
+		// Check if the availability is within the rehearsal time and if the start time is before the end time
+		const rehearsalStart = rehearsal.startTime.toDate()
+		const rehearsalEnd = rehearsal.endTime.toDate()
+
+		const start = startTimeString.split(':')
+		const end = endTimeString.split(':')
+
+		const availableStart = newTimeOnDay(rehearsalStart, parseInt(start[0]), parseInt(start[1]))
+		const availableEnd = newTimeOnDay(rehearsalEnd, parseInt(end[0]), parseInt(end[1]))
+
+		startTimeInvalid =
+			rehearsalStart > availableStart ||
+			rehearsalEnd < availableStart ||
+			availableStart >= availableEnd
+		endTimeInvalid =
+			rehearsalStart > availableEnd || rehearsalEnd < availableEnd || availableStart >= availableEnd
+
+		if (startTimeInvalid || endTimeInvalid) return false
+
+		// Save the availability to the database
+		const newAvailibilityDoc: NewAvailability = {
+			available,
+			rehearsal: doc(db, 'rehearsals/' + rehearsal.id),
+			reason: remark,
+			startTime: Timestamp.fromDate(availableStart),
+			endTime: Timestamp.fromDate(availableEnd)
+		}
+
+		const availabilityRef =
+			availability?.id !== undefined
+				? doc(db, 'users/' + user.id + '/availability/' + availability.id)
+				: doc(collection(db, 'users/' + user.id + '/availability'))
+
+		await setDoc(availabilityRef, newAvailibilityDoc)
+
+		invalidateAll()
+
+		return true
+	}
+
+	function toggleAvailable() {
+		available = !available
+		if (!available) {
+			startTimeString = getTimeString(rehearsal.startTime)
+			endTimeString = getTimeString(rehearsal.endTime)
+		}
+	}
+</script>
+
+<Modal
+	modalHeading={`${setOrUpdate} availability`}
+	bind:open={openSetAvailability}
+	primaryButtonText="Submit"
+	primaryButtonIcon={Edit}
+	secondaryButtonText="Cancel"
+	on:click:button--primary={async () => {
+		const success = await saveAvailability()
+		openSetAvailability = !success
+	}}
+	on:click:button--secondary={() => {
+		openSetAvailability = false
+	}}
+	on:close={() => onClose()}
+>
+	<Grid>
+		<Row padding>
+			<Column>
+				{`${setOrUpdate} ${
+					displayUserName ? user.name + "'s" : 'your'
```

Comment:
ok, so `displayUserName` is not actually `displayUserName`, but rather the negation of `isCurrentUser` or smth like that. If you name it like that it's clearer what to pass, then you can let the component sort out how to handle that